xamarin/xamarin-android
# [.NET 5] Resolve app content from runtime packs

Partially reworks the `BuildApk` task to accept a dynamic list of native
libraries needed for the app. System library names, and the logic which
would conditionally include them that was previously defined in the
`BuildApk` task has been moved into Xamarin.Android.Common.targets.

The `PreferNativeLibrariesWithDebugSymbols` parameter has been removed
from `BuildApk`.  We have not been shipping native libraries with debug
symbols, and as a result this check was not changing project packaging
behavior.

A RuntimeIdentifier <-> AndroidAbi mapping has been added to 
`MonoAndroidHelper.cs` to ensure that components coming from the 
`@(ResolvedFileToPublish)` ItemGroup would end up in the right location
in the .apk file.

The .NET 5 behavior for `_ResolveAssemblies` has been updated to also
use the `@(ResolvedFileToPublish)` items, to ensure that we include the
implementation version of these assemblies in our .apk / .aab files.

Linking has temporarily been disabled by default for .NET 5 apps while
a compatible solution is being worked on.

Shared runtime usage has also been disabled by default for .NET 5
apps.

AOT cross compilers and Android architecture specific native libraries
have been removed from Xamarin.Android.Sdk.

A couple of other runtime tweaks from
https://github.com/xamarin/xamarin-android/tree/test-net5-drop
have been applied to ensure a self published app can run.

### How to test

  1) Install the [latest .NET 5 preview][0].
  2) Download the .nupkg files from a CI build, and create a 
    `NuGet.config` file that has a package source pointing to them and
    the .NET 5 feed:

```xml
<configuration>
  <packageSources>
    <add key="xanetsdk" value="packages" />
    <add key="dotnet5" value="https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet5/nuget/v3/index.json" />
  </packageSources>
</configuration>
```

  3) Open an existing Android project (ideally something minimal) and
    tweak it as shown below. The version should match the version of the
    packages that you downloaded.

```xml
<Project Sdk="Xamarin.Android.Sdk/10.4.99.7">
  <PropertyGroup>
    <TargetFramework>netcoreapp5.0</TargetFramework>
    <RuntimeIdentifier>android.21-arm64</RuntimeIdentifier>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
  </PropertyGroup>
</Project>
```

  4) Publish (and optionally install) the project:

```
dotnet publish -t:Install -p:Configuration=Release *.csproj --self-contained
```

[0]:  https://github.com/dotnet/installer#installers-and-binaries
